### ISSUE TITLE
[ DestinE ] Allow to continue the workflow based on an specific status

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1024]
In GitLab by @dbeltrankyl on Apr 26, 2023, 12:48

All tasks should be able to continue the workflow depending on a list of statuses specified. 

* [ ] Add a new section to the DEPENDENCIES ( so we can do it to a task level) parameter to indicate from which status the workflow should progress.
* [ ] Add a way of polling this status (depending on the implementation of #1025))
* [ ] Add a way of recognizing it in Autosubmit
* [ ] Add setstatus -fcp? ( -fs is used for filter_status, -fcp would mean "filter_chunk_split)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 16, 2023, 10:29

Starting this one... 

Checking group_by & expand code

I want to emulate the status of a workflow with setstatus and.. realized that there is no filter for "splits" doing it with setatus -FL seems not viable when having more than +100.

**Comment by kinow** [comment]
In GitLab by @mcastril on May 31, 2023, 13:17

mentioned in issue digital-twins/de_340/workflow#25

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 6, 2023, 11:35

mentioned in issue digital-twins/de_340/project_management#118

**Comment by kinow** [comment]
In GitLab by @franra9 on Jun 21, 2023, 16:30

Hi @dbeltrankyl ,

If this is useful for you, in the applications workflow we would like to have a dependency on `status=RUNNING` so that we can launch the data notifier without having to wait for the model to finish.

Thanks :smile:

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 22, 2023, 09:47

Hello @franra9 

I have an issue with pickle in the #1025  issue, I think it will take a bit more, so I decided to start this one for 4.0 and then merge it to 4.1 so you can benefit from it ASAP, so all will be functional and the scalability/better visualization will come in 4.1, 

I'll explain what I will implement to achieve this issue in code, mostly for myself

Things to do: 

* [x] def update_list from job_list.py:
  * looks for the keys job.name and "checkpoint" in the edge_info dict
  * If it exists, retrieve the ending point of this edge and check if is running
  * If $state_to_check, looks for the file $(job.name)_step_$(AS_STEP_to_check)_completed

* [x] manage_dependencies function:
  * Looks in the yaml file if the special sign "!+"state(first letter?(S/Q/R))"+"step" is put in any dependency ( default could be just put "!"
  * Adds it to the edge_info dictionary with the key checkpoint,
  * Ensure that "?"(failed) and !(any_status) can be combined together? perhaps no need

* [x] Parameters function: 
  * Add %AS_CHECKPOINT% and %AS_STEP%=1 parameters to be used inside templates
  * Depending on the job.type:
    * Bash: touch %HPCROOT%/$job_name_step_$AS_STEP_completed # final command may change but similar
    * Python: (...)
    * R: (...)

* [x] Autosubmit yaml file
  * Adds in advanced dependency the sign "!" similar to "?" to the tasks/section that you wish

* [x] Workflow Developers template:
  * Add %AS_CHECKPOINT% whenever you want, 
  * ~~Increase %AS_STEP%+1 if multiple steps are needed~~ nvm This could be done integrated in the %AS_CHECKPOINT% one
  * Repeat if necessary

Edit: 
* Additional stuff: 

* [x] edge_info is currently at the job level. Add a dictionary to  the job_list and thus avoid iterating all running jobs and queuing and submitted ones ( I don't want to finish the performance issue and make a new one ) 

* [x] Manual tests and pytests

Fyi: @kinow

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 29, 2023, 11:40

As discussed in Autosubmit internal meeting, 

I will add a simpler version that doesn't needs to put anything in the templates, also I'll change the syntax

Before:

```yaml
Dependencies:
SIM:
  Splits_from:
   "1!r2":
      ... 
```

After:

```yaml
DEPENDENCIES:
SIM:
 STATUS: "SUBMITTED|READY|QUEUING|HOLD|FAILED|RUNNING|FAILED" # 
 FROM_STEP: "2" # Optional 
 SPLITS_FROM:
  1:
   ... 
```

If "FROM_STEP" is not found, or the STATUS not in ["FAILED","RUNNING"] THE %AS_CHECKPOINT% will be disabled ( as the update_list will enter in a different routine )

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 29, 2023, 11:46

@mcastril should I remove the mark "?" as you could just put STATUS: FAILED? 

Also, I forgot about the visualization, 

I thought about to the relation in this format "- - -" ( the same as when the "?" is put), and add colors that match the status to the line itself ( not sure if graphviz allows it I have to check )

**Comment by kinow** [comment]
In GitLab by @franra9 on Jun 29, 2023, 11:51

Hi @dbeltrankyl 

Is it contemplated that there is going to be a variable similar to `CHUNK_START_DATE`, `CHUNK_END_DATE` but for steps (i.e. `CHUNK_START_STEP`, `CHUNK_END_STEP`) (cc @ghosh97 @igonzal-bsc this is what you need for the request from FDB, right?)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 6, 2023, 09:00

In the end, I had to change most of the things in the code, but I think I'm closer to the final version. I also changed something to be more optimal with large workflows 

* ! has been removed. Instead, "?" changed its meaning to mark any conditional job, not only optional ones. 

* STATUS has to be declared in all cases
* If the status is declared, and there is no "?" , all dependencies will be conditional
* If the status is declared, and there is "?", only dependencies marked will be conditional
* FROM_STEP is optional. It only works for FAILED and RUNNING jobs ( as it has no sense for the other ones )
 
Current Syntax:  

```yaml

    DEPENDENCIES:
     SIM-1:
        STATUS: "RUNNING"
        FROM_STEP: "30" # optional
        SPLITS_FROM:
          '1':
            STATUS: "FAILED"
            SPLITS_TO: "1,2"
          '2':
            STATUS: "QUEUING"
            SPLITS_TO: '2,3'
```

The reason for the "?" mark.. is that if you want to depend on split_to: 1 completed, but you want to be split 2 conditional? It would help if you had a way to mark it.

* TODO: delete the old "?" code in favor of the new one ( STATUS: "FAILED" ) 
* TODO: Add colors to the - - - dashed
